### PR TITLE
fix typing for `Upload.beforeUpload`

### LINKED ISSUE #9775: Uploader component does not provide correct type info
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Version
3.3.1

### Environment
N/A

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/o7nl5x3kq5)

### Steps to reproduce
click upload, select a file

### What is expected?
console output `has originFileObj, is UploadFile!!!`

### What is actually happening?
console output `nop, not a upload file`

---
in Upload component's typing, it claims `beforeUpload` callback will receive a `UploadFile` typed object, but in fact we receive a `File` object.
Which causes a lot of trouble when using in TypeScript, for now we write a lot of `  as any as File` type cast just to get a correct type.

### PR DESCRIPTION
**Pull request by whtsky**
fix #9775

First of all, thank you for your contribution! :-)

Please makes sure that these checkboxes are checked before submitting your PR, thank you!

* [x] Make sure that you propose PR to right branch: bugfix for `master`, feature for latest active branch `feature-x.x`.
* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [x] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [x] Rebase before creating a PR to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you PR.

Extra checklist:

**if** *isBugFix* **:**

  * [x] Make sure that you add at least one unit test for the bug which you had fixed.

**elif** *isNewFeature* **:**

  * [x] Update API docs for the component.
  * [x] Update/Add demo to demonstrate new feature.
  * [x] Update TypeScript definition for the component.
  * [x] Add unit tests for the feature.

### CODE REVIEW

File: components/upload/__tests__/upload.test.js [hunk 1]
```diff
@@ -48,7 +48,7 @@ describe('Upload', () => {
     wrapper.find('input').simulate('change', {
       target: {
         files: [
-          { filename: 'foo.png' },
```
Comment (whtsky) [inline components/upload/__tests__/upload.test.js]: I looked up in [W3](https://www.w3.org/TR/FileAPI/#dfn-file) and [MDN](https://developer.mozilla.org/en-US/docs/Web/API/File), there's no doc about `filename` attr in File Object.

File: components/upload/Upload.tsx [hunk 2]
```diff
@@ -187,15 +188,15 @@ export default class Upload extends React.Component<UploadProps, UploadState> {
...
+  beforeUpload = (file: RcFile, fileList: RcFile[]) => {
     if (!this.props.beforeUpload) {
       return true;
     }
     const result = this.props.beforeUpload(file, fileList);
     if (result === false) {
       this.onChange({
-        file,
-        fileList: uniqBy(fileList.concat(this.state.fileList), (item: any) => item.uid),
+        file: fileToObject(file),
```
Comment (afc163) [inline components/upload/Upload.tsx]: #10293

Comment (whtsky) [reply components/upload/Upload.tsx]: typing 声称这里是 FileObject ，所以这是 bug fix 呀

Comment (whtsky) [reply components/upload/Upload.tsx]: typing 声称这里是 FileObject ，所以这是 bug fix 呀

Comment (afc163) [reply components/upload/Upload.tsx]: 主要会引起 break change。

File: components/upload/__tests__/upload.test.js (hunk 1 above)
Comment (JixunMoe) [reply components/upload/__tests__/upload.test.js]: 这个属性不应该是 `name` 吗…

https://developer.mozilla.org/en-US/docs/Web/API/File/name

**Pull request merged**